File: languages/python/sorting.py
import json
import os
import time
import statistics
import logging
from typing import List

import psutil

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)


def insertion_sort(arr: List[int]) -> List[int]:
    arr = arr.copy()
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr


def bubble_sort(arr: List[int]) -> List[int]:
    arr = arr.copy()
    n = len(arr)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break
    return arr

 
def load_test_data(size: int) -> List[int]:
    filename = f"data/test/test_data_{size}.json"
    if not os.path.exists(filename):
        raise FileNotFoundError(
            f"Test data file {filename} not found. Run generate_test_data.py first."
        )

    with open(filename, "r") as f:
        return json.load(f)


def measure_memory():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / 1024 / 1024


def benchmark_sorting_algorithm(algorithm, data: List[int], algorithm_name: str, iterations: int = 50):
    execution_times = []
    memory_used_values = []
    initial_memory_values = []
    final_memory_values = []

    logger.info(f"Starting benchmark for {algorithm_name} with {len(data)} elements ({iterations} iterations)")

    for i in range(iterations):
        initial_memory = measure_memory()

        start_time = time.time()

        algorithm(data)

        end_time = time.time()

        final_memory = measure_memory()

        execution_time = end_time - start_time
        memory_used = final_memory - initial_memory

        execution_times.append(execution_time)
        memory_used_values.append(memory_used)
        initial_memory_values.append(initial_memory)
        final_memory_values.append(final_memory)

        # Log every iteration completion
        logger.info(f"Completed iteration {i+1}/{iterations} for {algorithm_name} - Time: {execution_time:.6f}s, Memory: {memory_used:.2f}MB")

    return {
        "algorithm": algorithm_name,
        "data_size": len(data),
        "execution_time_mean": statistics.mean(execution_times),
        "execution_time_std": statistics.stdev(execution_times) if len(execution_times) > 1 else 0.0,
        "execution_time_min": min(execution_times),
        "execution_time_max": max(execution_times),
        "memory_used_mb_mean": statistics.mean(memory_used_values),
        "memory_used_mb_std": statistics.stdev(memory_used_values) if len(memory_used_values) > 1 else 0.0,
        "initial_memory_mb_mean": statistics.mean(initial_memory_values),
        "initial_memory_mb_std": statistics.stdev(initial_memory_values) if len(initial_memory_values) > 1 else 0.0,
        "final_memory_mb_mean": statistics.mean(final_memory_values),
        "final_memory_mb_std": statistics.stdev(final_memory_values) if len(final_memory_values) > 1 else 0.0,
        "iterations": iterations,
    }


def run_benchmarks():
    sizes = [1000, 5000, 10000]

    algorithms = [(insertion_sort, "Insertion Sort"), (bubble_sort, "Bubble Sort")]

    results = []

    logger.info(f"Starting benchmark suite with {len(sizes)} data sizes and {len(algorithms)} algorithms")

    for size_idx, size in enumerate(sizes):
        logger.info(f"Processing data size {size} ({size_idx+1}/{len(sizes)})")
        test_data = load_test_data(size)

        for alg_idx, (algorithm, name) in enumerate(algorithms):
            logger.info(f"Running {name} ({alg_idx+1}/{len(algorithms)}) on data size {size}")
            result = benchmark_sorting_algorithm(algorithm, test_data, name)
            results.append(result)
            logger.info(f"Completed {name} benchmark for size {size}")
    
    logger.info(f"Benchmark suite completed. Generated {len(results)} results.")
    return results


if __name__ == "__main__":
    logger.info("Starting Python sorting algorithm benchmark")
    results = run_benchmarks()

    with open("data/results/python_results.json", "w") as f:
        json.dump(results, f, indent=2)
    
    logger.info("Results saved to data/results/python_results.json")

File: languages/go/sorting.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"runtime"
	"time"
)

func InsertionSort(arr []int) []int {
	result := make([]int, len(arr))
	copy(result, arr)

	for i := 1; i < len(result); i++ {
		key := result[i]
		j := i - 1
		for j >= 0 && result[j] > key {
			result[j+1] = result[j]
			j--
		}
		result[j+1] = key
	}
	return result
}

func BubbleSort(arr []int) []int {
	result := make([]int, len(arr))
	copy(result, arr)

	n := len(result)
	for i := 0; i < n; i++ {
		swapped := false
		for j := 0; j < n-i-1; j++ {
			if result[j] > result[j+1] {
				result[j], result[j+1] = result[j+1], result[j]
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
	return result
}

func LoadTestData(size int) ([]int, error) {
	filename := fmt.Sprintf("data/test/test_data_%d.json", size)

	file, _ := os.Open(filename)

	defer file.Close()

	var data []int
	decoder := json.NewDecoder(file)
	decoder.Decode(&data)

	return data, nil
}

type BenchmarkResult struct {
	Algorithm           string  `json:"algorithm"`
	DataSize            int     `json:"data_size"`
	ExecutionTimeMean   float64 `json:"execution_time_mean"`
	ExecutionTimeStd    float64 `json:"execution_time_std"`
	ExecutionTimeMin    float64 `json:"execution_time_min"`
	ExecutionTimeMax    float64 `json:"execution_time_max"`
	MemoryUsedMBMean    float64 `json:"memory_used_mb_mean"`
	MemoryUsedMBStd     float64 `json:"memory_used_mb_std"`
	InitialMemoryMBMean float64 `json:"initial_memory_mb_mean"`
	InitialMemoryMBStd  float64 `json:"initial_memory_mb_std"`
	FinalMemoryMBMean   float64 `json:"final_memory_mb_mean"`
	FinalMemoryMBStd    float64 `json:"final_memory_mb_std"`
	Iterations          int     `json:"iterations"`
}

func MeasureMemory() float64 {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	return float64(m.Alloc) / 1024 / 1024
}

func calculateMean(values []float64) float64 {
	if len(values) == 0 {
		return 0.0
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	return sum / float64(len(values))
}

func calculateStdDev(values []float64) float64 {
	if len(values) <= 1 {
		return 0.0
	}
	mean := calculateMean(values)
	sumSquaredDiffs := 0.0
	for _, v := range values {
		sumSquaredDiffs += math.Pow(v-mean, 2)
	}
	return math.Sqrt(sumSquaredDiffs / float64(len(values)-1))
}

func calculateMin(values []float64) float64 {
	if len(values) == 0 {
		return 0.0
	}
	min := values[0]
	for _, v := range values {
		if v < min {
			min = v
		}
	}
	return min
}

func calculateMax(values []float64) float64 {
	if len(values) == 0 {
		return 0.0
	}
	max := values[0]
	for _, v := range values {
		if v > max {
			max = v
		}
	}
	return max
}

func BenchmarkSortingAlgorithm(algorithm func([]int) []int, data []int, algorithmName string, iterations int) BenchmarkResult {
	var executionTimes []float64
	var memoryUsedValues []float64
	var initialMemoryValues []float64
	var finalMemoryValues []float64

	log.Printf("Starting benchmark for %s with %d elements (%d iterations)",
		algorithmName, len(data), iterations)

	for i := 0; i < iterations; i++ {
		runtime.GC()

		initialMemory := MeasureMemory()
		startTime := time.Now()

		algorithm(data)

		endTime := time.Now()
		finalMemory := MeasureMemory()

		executionTime := endTime.Sub(startTime)
		memoryUsed := finalMemory - initialMemory

		executionTimes = append(executionTimes, executionTime.Seconds())
		memoryUsedValues = append(memoryUsedValues, memoryUsed)
		initialMemoryValues = append(initialMemoryValues, initialMemory)
		finalMemoryValues = append(finalMemoryValues, finalMemory)

		// Log every iteration completion
		log.Printf("Completed iteration %d/%d for %s - Time: %.6fs, Memory: %.2fMB",
			i+1, iterations, algorithmName, executionTime.Seconds(), memoryUsed)
	}

	return BenchmarkResult{
		Algorithm:           algorithmName,
		DataSize:            len(data),
		ExecutionTimeMean:   calculateMean(executionTimes),
		ExecutionTimeStd:    calculateStdDev(executionTimes),
		ExecutionTimeMin:    calculateMin(executionTimes),
		ExecutionTimeMax:    calculateMax(executionTimes),
		MemoryUsedMBMean:    calculateMean(memoryUsedValues),
		MemoryUsedMBStd:     calculateStdDev(memoryUsedValues),
		InitialMemoryMBMean: calculateMean(initialMemoryValues),
		InitialMemoryMBStd:  calculateStdDev(initialMemoryValues),
		FinalMemoryMBMean:   calculateMean(finalMemoryValues),
		FinalMemoryMBStd:    calculateStdDev(finalMemoryValues),
		Iterations:          iterations,
	}
}

func RunBenchmarks() []BenchmarkResult {
	sizes := []int{1000, 5000, 10000}

	algorithms := []struct {
		Function func([]int) []int
		Name     string
	}{
		{InsertionSort, "Insertion Sort"},
		{BubbleSort, "Bubble Sort"},
	}

	var results []BenchmarkResult

	log.Printf("Starting benchmark suite with %d data sizes and %d algorithms",
		len(sizes), len(algorithms))

	for sizeIdx, size := range sizes {
		log.Printf("Processing data size %d (%d/%d)", size, sizeIdx+1, len(sizes))
		testData, _ := LoadTestData(size)

		for algIdx, alg := range algorithms {
			log.Printf("Running %s (%d/%d) on data size %d",
				alg.Name, algIdx+1, len(algorithms), size)
			result := BenchmarkSortingAlgorithm(alg.Function, testData, alg.Name, 50)
			results = append(results, result)
			log.Printf("Completed %s benchmark for size %d", alg.Name, size)
		}
	}

	log.Printf("Benchmark suite completed. Generated %d results.", len(results))
	return results
}

func main() {
	log.Printf("Starting Go sorting algorithm benchmark")
	rand.Seed(time.Now().UnixNano())

	results := RunBenchmarks()

	jsonFile, _ := os.Create("data/results/go_results.json")
	defer jsonFile.Close()

	encoder := json.NewEncoder(jsonFile)
	encoder.SetIndent("", "  ")
	encoder.Encode(results)

	log.Printf("Results saved to data/results/go_results.json")
}
